### ISSUE TITLE
indexThreadQty_  is uninitialized if index is loaded but not created (SmallWorldRand)

### ISSUE THREAD

**Issue by BorisLestsov** [nmslib/nmslib#351]
As far as I see, the only place where indexThreadQty_ gets initialized is SmallWorldRand::CreateIndex.
If index is loaded by SmallWorldRand::LoadIndex and not created by SmallWorldRand::CreateIndex, errors might occur.

For example, error occurs in "/similarity_search/src/method/small_world_rand.cc" if addBatch is called (throws std::bad_alloc).

![image](https://user-images.githubusercontent.com/17254945/47110069-8d0a0800-d258-11e8-90f5-fe5f9b699853.png)

**Comment by searchivarius** [comment]
@BorisLestsov ohhh, right, b/c mostly we don't add anything after the index is created. The only mode is an experimental app that can add and delete the stuff incrementally. So, anyways, this is an easy thing to fix. Would you care to submit a pull request? The number of threads should be deduced from a respective index-time parameter. 

Thank you!

**Comment by BorisLestsov** [comment]
@searchivarius Thank you for your answer. I might submit a pull request if I will manage adding and deleting objects from my index :D Right now I am kind of confused, so could you provide some example of how to use this functionality? Or at least describe what should I do to add and delete elements from loaded index. I have not found anything about this in docs.

Right now, I imagine the whole process this way:
1) I add vectors to index through SmallWorldRand constructor.
2) I call CreateIndex
3) I call SaveIndex

Then, when I want to load Index and add new elements I do:
1)  Add vectors to index through SmallWorldRand constructor the same way.
2) I call LoadIndex
3) I call addBatch or DeleteBatch, index automatically completed for new elements or patched in case if deletion.

Is this right reasoning?

**Comment by searchivarius** [comment]
@BorisLestsov sorry for the delay, the testing app is here: https://github.com/nmslib/nmslib/tree/master/test_batch_app
It is all very experimental. After reading through the code I have doubts you can save a dynamic index. Probably not.

**Comment by BorisLestsov** [comment]
@searchivarius 
Thank you for your answer. Actually, after some code modification a was able to save Index after elements were added and it works correct.
BTW, I think I found a critical bug in SmallWorldRand::AddBatch. If I am right I will make pull request later. The problem is here: https://github.com/nmslib/nmslib/blob/master/similarity_search/src/method/small_world_rand.cc#L154
Basically, this line means, that 0-th element of the batch is only added to the node vector and is not linked to any previous nodes. It works correct when index is empty (i.e. AddBatch is called for the first time), but any successive calls of AddBatch might lead to incorrect results.
